# Diverse secondary metabolites are expressed in particle-associated and free-living microorganisms of the permanently anoxic Cariaco Basin

[![DOI](https://zenodo.org/badge/509871897.svg)](https://zenodo.org/badge/latestdoi/509871897)

Scripts for analysis of metagenomic and metatranscriptomic datasets from the Cariaco Basin, Venezuela. To view the code used for this analysis, click here: https://d-mcgrath.github.io/cariaco_basin/. This is a GitHub pages document that contains R scripts for figure creation and data analysis, as well as bash scripts for data analysis that was done using a high performance computing cluster.

Data required to re-run the analyses are available in the [cariaco_github_data_small](https://github.com/d-mcgrath/cariaco_basin/tree/main/cariaco_github_data_small) folder. Three data files that were too large to host on GitHub are available in an OSF repository in the folder "Large GitHub files" from the following link: https://osf.io/usm8r/. Several Conda environments required to reproduce these analyses are in the [yaml_files_for_analysis](https://github.com/d-mcgrath/cariaco_basin/tree/main/yaml_files_for_analysis) folder.

All MAGs recovered in this study and biosynthetic gene clusters recovered from those MAGs using antiSMASH 6.0 are available in the repository: https://osf.io/usm8r/ in the folders "Prokaryotic_CarAnox_MAGs" and "Biosynthetic_gene_clusters". All data in .tsv format used to create the figures and supplementary figures in this study are available in the [source_files](https://github.com/d-mcgrath/cariaco_basin/tree/main/source_files) folder. The supplementary tables from the manuscript can be found in the [supplementary_data](https://github.com/d-mcgrath/cariaco_basin/tree/main/supplementary_data) folder.
