# Diverse secondary metabolites are expressed in particle-associated and free-living microorganisms of the permanently anoxic Cariaco Basin

[![DOI](https://zenodo.org/badge/509871897.svg)](https://zenodo.org/badge/latestdoi/509871897)

This repository contains scripts, data, and computing environments used for the analysis of metagenomic and metatranscriptomic datasets from the Cariaco Basin, Venezuela for the publication "Diverse secondary metabolites are expressed in particle-associated and free-living microorganisms of the permanently anoxic Cariaco Basin". The authors of this paper are D. Geller-McGrath, P. Mara, G. Taylor, E. Suter, V. Edgcomb, and M. Pachiadaki. The scripts, files, and data here were written and uploaded by D. Geller-McGrath.

To view the code used for this analysis, click here: https://d-mcgrath.github.io/cariaco_basin/. This is a GitHub pages document that contains R scripts for figure creation and data analysis, as well as bash scripts for data analysis that was done using a high performance computing cluster (HPC). The R script chunks contain in order the code for creation of figures 1-4 in the main text, and supplementary figures 1-6. Bash scripts for analysis done on an HPC are below the R scripts. Several Conda environments required to reproduce these HPC analyses are in the [yaml_files_for_analysis](https://github.com/d-mcgrath/cariaco_basin/tree/main/yaml_files_for_analysis) folder. Data required to re-run the R scripts are available in the [cariaco_github_data_small](https://github.com/d-mcgrath/cariaco_basin/tree/main/cariaco_github_data_small) folder. Four data files that were too large (total size < 1G) to host on GitHub are available in an OSF repository in the folder "Large GitHub files" from the following link: https://osf.io/usm8r/. 

All MAGs recovered in this study and biosynthetic gene clusters recovered from those MAGs using antiSMASH 6.0 are available in the repository: https://osf.io/usm8r/ in the folders "Prokaryotic_CarAnox_MAGs" and "Biosynthetic_gene_clusters". All data in .tsv format used to create the figures and supplementary figures in this study are available in the [source_files](https://github.com/d-mcgrath/cariaco_basin/tree/main/source_files) folder. These source files can be loaded into an R environment using the `readRDS()` function, and the corresponding ggplot2 code for figure generation can be used directly with these R objects. The supplementary data (in an excel workbook) from the manuscript can be found in the [supplementary_data](https://github.com/d-mcgrath/cariaco_basin/tree/main/supplementary_data) folder. The raw metagenomic and metatranscriptomic sequencing reads used in this study can be accessed on [NCBI](https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA326482).
